File: src/py/agent_studio/core/paths.py
"""Resolve project/config/data directories.

Mirrors the TypeScript ``resolvePaths`` so both runtimes read and write the same
locations. All paths are overridable via environment variables.
"""

from __future__ import annotations

import os
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path


def find_project_root(start: Path) -> Path:
    """Walk upward from ``start`` to the nearest directory containing
    ``package.json`` (the repo root marker used by the TS side)."""
    d = start.resolve()
    for _ in range(20):
        if (d / "package.json").exists():
            return d
        if d.parent == d:
            break
        d = d.parent
    return start.resolve()


@dataclass(frozen=True)
class ResolvedPaths:
    project_root: Path
    config_dir: Path
    data_dir: Path
    settings_file: Path


def resolve_paths(env: Mapping[str, str] | None = None) -> ResolvedPaths:
    env = os.environ if env is None else env
    root = find_project_root(Path(__file__).parent)

    config_override = env.get("AGENT_STUDIO_CONFIG_DIR")
    data_override = env.get("AGENT_STUDIO_DATA_DIR")

    config_dir = Path(config_override).resolve() if config_override else root / "config"
    data_dir = Path(data_override).resolve() if data_override else root / ".agentstudio"
    return ResolvedPaths(root, config_dir, data_dir, data_dir / "settings.json")

File: src/engine/core/paths.ts
// Resolves project/config/data directories. All locations are overridable via
// environment variables so users can relocate config/data without code changes.

import { existsSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';

function findProjectRoot(startDir: string): string {
  let dir = startDir;
  // Walk upward until we find a package.json (project root marker).
  for (let i = 0; i < 20; i++) {
    if (existsSync(join(dir, 'package.json'))) return dir;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return startDir;
}

export interface ResolvedPaths {
  projectRoot: string;
  configDir: string;
  dataDir: string;
  settingsFile: string;
}

export function resolvePaths(env: NodeJS.ProcessEnv = process.env): ResolvedPaths {
  const projectRoot = findProjectRoot(import.meta.dirname);
  const configDir = env.AGENT_STUDIO_CONFIG_DIR
    ? resolve(env.AGENT_STUDIO_CONFIG_DIR)
    : join(projectRoot, 'config');
  const dataDir = env.AGENT_STUDIO_DATA_DIR
    ? resolve(env.AGENT_STUDIO_DATA_DIR)
    : join(projectRoot, '.agentstudio');
  const settingsFile = join(dataDir, 'settings.json');
  return { projectRoot, configDir, dataDir, settingsFile };
}
